File: problems/sorting/LC189_rotate_array/solution.cpp
#include <iostream>
#include <vector>
#include <algorithm>

// space-efficient
class Solution1 {
public:
  void rotate(std::vector<int>& nums, int k) {
    std::reverse(nums.end() - k, nums.end());
    std::reverse(nums.begin(), nums.end() - k);
    std::reverse(nums.begin(), nums.end());
  }
};

// time-efficient
class Solution2 {
public:
  void rotate(std::vector<int>& nums, int k) {
    if (nums.size() == 1) {
      return;
    }
    while (k > nums.size()) {
      k -= nums.size();
    }
    for (int i = 0; i < nums.size() - k; i++) {
      nums.push_back(nums[i]);
    }
    nums.erase(nums.begin() + 0,nums.begin() + nums.size() - k);
  }
};

int main() {
  std::vector<int> nums = {1,2,3,4,5,6,7};
  int k = 3;
  Solution1 s;
  s.rotate(nums, k);
  for (int num: nums) {
    std::cout << num << std::endl;
  }
}

File: problems/sorting/LC189_rotate_array/solution.go
package main

import (
	"fmt"
	"strconv"
)

// reverse array helper function
func reverse(nums []int, start, end int) {
	for start < end {
		nums[start], nums[end] = nums[end], nums[start]
		start++
		end--
	}
}

// Solution 1: space-efficient
func rotate1(nums []int, k int) {
	n := len(nums)
	reverse(nums, n-k, n-1)
	reverse(nums, 0, n-k-1)
	reverse(nums, 0, n-1)
}

// Solution 2: time-efficient
func rotate2(nums []int, k int) {
	n := len(nums)
	k = k % n

	if n == 1 {
		return
	}

	for i := 0; i < n-k; i++ {
		nums = append(nums, nums[i])
	}

	nums = nums[n-k:]
}

func main() {
	nums := []int{1, 2, 3, 4, 5, 6, 7}
	k := 3
	rotate1(nums, k)
	for i := 0; i < len(nums); i++ {
		fmt.Print(strconv.Itoa(nums[i]) + " ")
	}
}
